File: rust/389.find-the-difference.rs
/*
 * @lc app=leetcode id=389 lang=rust
 *
 * [389] Find the Difference
 */

// @lc code=start
impl Solution {
    pub fn find_the_difference(s: String, t: String) -> char {
        // Solution 1:
        // let mut chars = vec![0; 26];
        // for c in t.bytes() {
        //     chars[(c - b'a') as usize] += 1;
        // }
        // for c in s.bytes() {
        //     chars[(c - b'a') as usize] -= 1;
        // }
        // return (chars.iter().position(|c| *c == 1).unwrap() as u8 + b'a') as char;

        // Solution 2: using XOR
        let mut diff = 0;
        for c in t.bytes() {
            diff ^= c;
        }
        for c in s.bytes() {
            diff ^= c;
        }
        return diff as char;
    }
}
// @lc code=end

File: javascript/389.find-the-difference.js
/*
 * @lc app=leetcode id=389 lang=javascript
 *
 * [389] Find the Difference
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} t
 * @return {character}
 */
var findTheDifference = function (s, t) {
  // Solution 1:
  // const chars = new Array(26).fill(0);
  // const d = "a".charCodeAt();
  // for (const c of t) chars[c.charCodeAt() - d]++;
  // for (const c of s) chars[c.charCodeAt() - d]--;
  // return String.fromCharCode(chars.findIndex((c) => c === 1) + d);

  // Solution 2: using XOR
  let diff = 0;
  for (const c of s) diff ^= c.charCodeAt();
  for (const c of t) diff ^= c.charCodeAt();
  return String.fromCharCode(diff);
};
// @lc code=end
